boa/src/builtins/json/tests.rs

```diff
@@ -16,3 +16,155 @@ fn json_sanity() {
+    );
+    let expected = forward(&mut engine, r#"'{"aaa":"bbb","bbb":"ccc"}'"#);
+    assert_eq!(actual, expected);
+}
+
+#[test]
+fn json_stringify_replacer_array_numbers() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+    let actual = forward(
+        &mut engine,
+        r#"JSON.stringify({ 0: 'aaa', 1: 'bbb', 2: 'ccc'}, [1, 2])"#,
+    );
+    let expected = forward(&mut engine, r#"'{"1":"bbb","2":"ccc"}'"#);
+    assert_eq!(actual, expected);
+}
+
+#[test]
+fn json_stringify_replacer_function() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+    let actual = forward(
+        &mut engine,
+        r#"JSON.stringify({ aaa: 1, bbb: 2}, (key, value) => {
+            if (key === 'aaa') {
+                return undefined;
+            }
+
+            return value;
+        })"#,
+    );
+    let expected = forward(&mut engine, r#"'{"bbb":2}'"#);
+    assert_eq!(actual, expected);
+}
+
+#[test]
+fn json_stringify_arrays() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+    let actual = forward(&mut engine, r#"JSON.stringify(['a', 'b'])"#);
+    let expected = forward(&mut engine, r#"'["a","b"]'"#);
+
+    assert_eq!(actual, expected);
+}
+
+#[test]
+fn json_stringify_object_array() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+    let actual = forward(&mut engine, r#"JSON.stringify([{a: 'b'}, {b: 'c'}])"#);
+    let expected = forward(&mut engine, r#"'[{"a":"b"},{"b":"c"}]'"#);
+
+    assert_eq!(actual, expected);
+}
+
+#[test]
+fn json_stringify_array_converts_undefined_to_null() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+    let actual = forward(&mut engine, r#"JSON.stringify([undefined])"#);
+    let expected = forward(&mut engine, r#"'[null]'"#);
+
+    assert_eq!(actual, expected);
+}
+
+#[test]
+fn json_stringify_array_converts_function_to_null() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+    let actual = forward(&mut engine, r#"JSON.stringify([() => {}])"#);
+    let expected = forward(&mut engine, r#"'[null]'"#);
+
+    assert_eq!(actual, expected);
+}
+
+#[test]
+#[ignore]
+fn json_stringify_array_converts_symbol_to_null() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+    let actual = forward(&mut engine, r#"JSON.stringify([Symbol()])"#);
+    let expected = forward(&mut engine, r#"'[null]'"#);
+
+    assert_eq!(actual, expected);
+}
+#[test]
+fn json_stringify_function_replacer_propogate_error() {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let actual = forward(
+        &mut engine,
+        r#"JSON.stringify({x: 1}, (key, value) => {
+         throw new TypeError("type error")
+     })"#,
+    );
+    let expected = forward(&mut engine, r#"'Error: undefined'"#);
+
+    assert_eq!(actual, expected);
+}
```

Comment:
Not entirely sure how to test this. I followed the same pattern that has been applied in this file and just captured the output of what I think you were suggesting in the production code.

Reply:
We don't have a `TypeError` yet, so that's why this is returning `undefined`. What we could do is to just put it inside of a `try..catch`:

```javascript
	let a = 0;
    try {
		JSON.stringify({x: 1}, (key, value) => {
			throw 1;
    	});
	} catch (e) {
		a = e;
	}
	a;
```

This should output `1`.

Reply:
@Razican -- Done :heavy_check_mark: